File: frontend/src/api/teacher.ts
/**
 * api/teacher.ts
 * ==============
 * Clientes HTTP tipados para los endpoints de docente y admin.
 */

import { api } from "./client";

const API_BASE = import.meta.env.VITE_API_URL ?? "";

// ── Tipos compartidos ─────────────────────────────────────────────────────────

export interface Group {
  group_id: number;
  name: string;
  course_id: string | null;
  invite_code: string | null;
  student_count: number;
}

export interface StudentSummary {
  user_id: number;
  username: string;
  global_elo: number;
  total_attempts: number;
  accuracy: number;
  last_activity: string | null;
  group_id?: number | null;
  group_name?: string | null;
  education_level?: string | null;
}

export interface DashboardData {
  teacher_id: number;
  groups: Group[];
  students: StudentSummary[];
}

export interface TopicStat {
  topic: string;
  attempts: number;
  accuracy: number;
  avg_time: number;
}

export interface TeacherMetrics {
  total_attempts: number;
  avg_time_seconds: number;
  abandonment_rate: number;
  topic_stats: TopicStat[];
  daily_attempts: { date: string; count: number }[];
  hourly_distribution: { hour: number; count: number }[];
}

export interface PendingProcedure {
  submission_id: number;
  student_id: number;
  student_username: string;
  item_id: string;
  item_content: string | null;
  ai_score: number | null;
  status: string;
  created_at: string;
  has_image: boolean;
}

export interface GradeResult {
  submission_id: number;
  teacher_score: number;
  elo_delta: number;
  status: string;
}

export interface AdminUser {
  id?: number;
  user_id?: number;
  username: string;
  role: string;
  approved: boolean | number;
  active: boolean | number;
  education_level: string | null;
  group_name: string | null;
}

export interface AdminGroup {
  id: number;
  name: string;
  teacher_id?: number;
  teacher_username?: string;
  student_count?: number;
}

export interface ProblemReport {
  id: number;
  user_id: number;
  username?: string;
  description: string;
  status: string;
  created_at: string;
}

export interface AuditEntry {
  id: number;
  student_id: number;
  student_username?: string;
  old_group_id: number | null;
  old_group_name: string | null;
  new_group_id: number | null;
  new_group_name: string | null;
  admin_id: number;
  admin_username?: string;
  timestamp: string;
}

// ── Descarga de archivos (blob) ───────────────────────────────────────────────

async function _downloadBlob(path: string, filename: string): Promise<void> {
  const { useAuthStore } = await import("../stores/authStore");
  const token = useAuthStore.getState().accessToken;
  const res = await fetch(`${API_BASE}${path}`, {
    headers: token ? { Authorization: `Bearer ${token}` } : {},
    credentials: "include",
  });
  if (!res.ok) throw new Error(`Error al descargar: HTTP ${res.status}`);
  const blob = await res.blob();
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = filename;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
}

// ── API Docente ───────────────────────────────────────────────────────────────

export const teacherApi = {
  dashboard: () => api.get<DashboardData>("/api/teacher/dashboard"),

  groups: () => api.get<Group[]>("/api/teacher/groups"),

  createGroup: (course_id: string, group_name: string) =>
    api.post<Group>("/api/teacher/groups", { course_id, group_name }),

  generateInviteCode: (group_id: number) =>
    api.post<{ invite_code: string }>(`/api/teacher/groups/${group_id}/invite-code`),

  procedures: () => api.get<PendingProcedure[]>("/api/teacher/procedures"),

  gradeProcedure: (submission_id: number, teacher_score: number, teacher_feedback?: string) =>
    api.post<GradeResult>("/api/teacher/procedures/grade", {
      submission_id,
      teacher_score,
      teacher_feedback,
    }),

  studentReport: (student_id: number) =>
    api.get<Record<string, unknown>>(`/api/teacher/student/${student_id}`),

  studentEloHistory: (student_id: number) =>
    api.get<{ attempts: Record<string, unknown>[] }>(`/api/teacher/student/${student_id}/elo-history`),

  studentKatiaHistory: (student_id: number) =>
    api.get<{ interactions: Record<string, unknown>[] }>(
      `/api/teacher/student/${student_id}/katia-history`,
    ),

  studentAiAnalysis: (student_id: number, api_key?: string, provider = "groq") =>
    api.post<{ analysis: string }>(`/api/teacher/student/${student_id}/ai-analysis`, {
      api_key,
      provider,
    }),

  procedureImage: async (submission_id: number): Promise<string> => {
    const { useAuthStore } = await import("../stores/authStore");
    const token = useAuthStore.getState().accessToken;
    const r = await fetch(`${API_BASE}/api/teacher/procedures/${submission_id}/image`, {
      headers: token ? { Authorization: `Bearer ${token}` } : {},
      credentials: "include",
    });
    if (!r.ok) throw new Error(`HTTP ${r.status}`);
    const blob = await r.blob();
    return URL.createObjectURL(blob);
  },

  metrics: () => api.get<TeacherMetrics>("/api/teacher/metrics"),

  downloadCsv: () => _downloadBlob("/api/teacher/export/csv", "levelup_intentos.csv"),

  downloadXlsx: () => _downloadBlob("/api/teacher/export/xlsx", "levelup_datos_completos.xlsx"),

  // ── Sprint C: plantillas de examen del docente ───────────────────────────
  allCourses: () => api.get<{ id: string; name: string; block: string }[]>("/api/teacher/courses"),

  examTemplates: (course_id?: string, include_archived = false) => {
    const qs = new URLSearchParams();
    if (course_id) qs.set("course_id", course_id);
    if (include_archived) qs.set("include_archived", "true");
    return api.get<ExamTemplate[]>(
      `/api/teacher/exam-templates${qs.toString() ? "?" + qs.toString() : ""}`,
    );
  },

  createExamTemplate: (input: {
    course_id: string;
    title: string;
    time_limit_min: number;
    item_ids: string[];
  }) => api.post<ExamTemplate>("/api/teacher/exam-templates", input),

  updateExamTemplate: (
    id: number,
    patch: {
      title?: string;
      time_limit_min?: number;
      item_ids?: string[];
    },
  ) => api.patch<ExamTemplate>(`/api/teacher/exam-templates/${id}`, patch),

  archiveExamTemplate: (id: number) =>
    api.delete<void>(`/api/teacher/exam-templates/${id}`),

  itemsCatalog: (course_id: string) =>
    api.get<ItemCatalogEntry[]>(
      `/api/teacher/items?course_id=${encodeURIComponent(course_id)}`,
    ),

  // ── Asignación de plantillas a grupos + ventana de tiempo ──────────────
  listAssignments: (template_id: number) =>
    api.get<ExamAssignment[]>(`/api/teacher/exam-templates/${template_id}/assignments`),

  createAssignments: (
    template_id: number,
    body: { group_ids: number[]; starts_at?: string | null; ends_at?: string | null },
  ) =>
    api.post<ExamAssignment[]>(
      `/api/teacher/exam-templates/${template_id}/assignments`,
      body,
    ),

  deleteAssignment: (template_id: number, assignment_id: number) =>
    api.delete<void>(
      `/api/teacher/exam-templates/${template_id}/assignments/${assignment_id}`,
    ),

  examResults: (template_id: number) =>
    api.get<ExamResults>(`/api/teacher/exam-templates/${template_id}/results`),
};

export interface ExamQuestionStat {
  item_id: string;
  content: string;
  topic: string | null;
  total: number;
  correct: number;
  accuracy: number;
}
export interface ExamTopicStat {
  topic: string;
  total: number;
  correct: number;
  accuracy: number;
}
export interface ExamResults {
  title: string;
  n_sessions: number;
  n_students: number;
  avg_score: number;
  questions: ExamQuestionStat[];
  topics: ExamTopicStat[];
  best_question: ExamQuestionStat | null;
  worst_question: ExamQuestionStat | null;
  reinforce_topic: ExamTopicStat | null;
}

export interface ExamAssignment {
  id: number;
  template_id: number;
  group_id: number;
  group_name: string;
  starts_at: string | null;
  ends_at: string | null;
  created_at: string;
}

export interface ExamTemplate {
  id: number;
  teacher_id: number;
  course_id: string;
  title: string;
  time_limit_min: number;
  item_ids: string[];
  archived: boolean;
  created_at: string;
}

export interface ItemCatalogEntry {
  id: string;
  content: string;
  difficulty: number;
  topic: string;
  tags: string[];
}

// ── API Admin ─────────────────────────────────────────────────────────────────

export const adminApi = {
  users: () => api.get<{ users: AdminUser[] }>("/api/admin/users"),

  pendingTeachers: () => api.get<{ teachers: AdminUser[] }>("/api/admin/teachers/pending"),

  approveTeacher: (user_id: number, action: "approve" | "reject") =>
    api.post<{ message: string }>("/api/admin/teachers/approve", { user_id, action }),

  deactivate: (user_id: number) =>
    api.patch<void>(`/api/admin/users/${user_id}/deactivate`),

  reactivate: (user_id: number) =>
    api.patch<void>(`/api/admin/users/${user_id}/reactivate`),

  changeGroup: (student_id: number, new_group_id: number | null) =>
    api.patch<void>("/api/admin/students/group", { student_id, new_group_id }),

  allGroups: () => api.get<{ groups: AdminGroup[] }>("/api/admin/groups"),

  deleteGroup: (group_id: number) => api.delete<void>(`/api/admin/groups/${group_id}`),

  reports: () => api.get<{ reports: ProblemReport[] }>("/api/admin/reports"),

  resolveReport: (report_id: number) =>
    api.patch<void>(`/api/admin/reports/${report_id}/resolve`),

  audit: (limit: number = 100) =>
    api.get<{ entries: AuditEntry[] }>(`/api/admin/audit?limit=${limit}`),
};

File: api/schemas/teacher.py
"""
api/schemas/teacher.py
======================
Pydantic schemas para los endpoints del docente y admin.
"""

from pydantic import BaseModel, Field


# ── Grupos ────────────────────────────────────────────────────────────────────


class CreateGroupRequest(BaseModel):
    course_id: str
    group_name: str = Field(..., min_length=1, max_length=80)


class GroupResponse(BaseModel):
    group_id: int
    name: str
    course_id: str | None
    invite_code: str | None
    student_count: int = 0


# ── Dashboard ─────────────────────────────────────────────────────────────────


class StudentSummary(BaseModel):
    user_id: int
    username: str
    global_elo: float
    total_attempts: int
    accuracy: float  # 0.0–1.0
    last_activity: str | None
    # Campos para filtros cascada (opcionales para compatibilidad)
    group_id: int | None = None
    group_name: str | None = None
    education_level: str | None = None


class DashboardResponse(BaseModel):
    teacher_id: int
    groups: list[GroupResponse]
    students: list[StudentSummary]


# ── Procedimientos ────────────────────────────────────────────────────────────


class GradeRequest(BaseModel):
    submission_id: int
    teacher_score: float = Field(..., ge=0, le=100)
    teacher_feedback: str | None = None


class GradeResponse(BaseModel):
    submission_id: int
    teacher_score: float
    elo_delta: float
    status: str


class PendingProcedure(BaseModel):
    submission_id: int
    student_id: int
    student_username: str
    item_id: str
    item_content: str | None
    ai_score: float | None
    status: str
    created_at: str
    has_image: bool = False  # True si hay imagen disponible para ver


# ── Reporte por estudiante ────────────────────────────────────────────────────


class StudentReportResponse(BaseModel):
    user_id: int
    username: str
    global_elo: float
    topic_breakdown: dict  # topic → {rating, rd, attempts}
    recent_attempts: list[dict]
    procedure_history: list[dict]


# ── Admin ─────────────────────────────────────────────────────────────────────


class UserAdminRow(BaseModel):
    user_id: int
    username: str
    role: str
    approved: bool
    active: bool
    education_level: str | None
    group_name: str | None


class ApproveTeacherRequest(BaseModel):
    user_id: int
    action: str = Field(..., pattern="^(approve|reject)$")


class ChangeGroupRequest(BaseModel):
    student_id: int
    new_group_id: int | None


# ── Sprint C: Plantillas de examen del docente ───────────────────────────────


class ExamTemplateCreateRequest(BaseModel):
    course_id: str = Field(..., min_length=1)
    title: str = Field(..., min_length=1, max_length=140)
    time_limit_min: int = Field(default=20, ge=5, le=180)
    item_ids: list[str] = Field(..., min_length=1, max_length=60)


class ExamTemplatePatchRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=140)
    time_limit_min: int | None = Field(default=None, ge=5, le=180)
    item_ids: list[str] | None = Field(default=None, min_length=1, max_length=60)


class ExamTemplateResponse(BaseModel):
    id: int
    teacher_id: int
    course_id: str
    title: str
    time_limit_min: int
    item_ids: list[str]
    archived: bool
    created_at: str


class ExamAssignmentCreateRequest(BaseModel):
    """Asigna una plantilla a uno o varios grupos con ventana de tiempo opcional."""

    group_ids: list[int] = Field(..., min_length=1, max_length=50)
    starts_at: str | None = Field(
        default=None,
        description="ISO datetime UTC. None = abierto desde ya.",
    )
    ends_at: str | None = Field(
        default=None,
        description="ISO datetime UTC. None = sin fecha de cierre.",
    )


class ExamAssignmentResponse(BaseModel):
    id: int
    template_id: int
    group_id: int
    group_name: str
    starts_at: str | None
    ends_at: str | None
    created_at: str


class ItemCatalogEntry(BaseModel):
    """Resumen de item para que el docente lo seleccione al armar el examen."""

    id: str
    content: str
    difficulty: float
    topic: str
    tags: list[str] = []
